src/datadog_context.cpp

```diff
@@ -32,12 +41,47 @@ void DatadogContext::on_change_block(ngx_http_request_t *request,
                        &traces_[0].active_span());
 }
 
+#ifdef WITH_WAF
+bool DatadogContext::on_main_req_access(ngx_http_request_t *request) {
+  if (!sec_ctx_) {
+    return false;
+  }
+
+  // there should only one trace at this point
+  dd::Span &span = single_trace().active_span();
+  return sec_ctx_->on_request_start(*request, span);
+}
+#endif
+
+#ifdef WITH_WAF
+ngx_int_t DatadogContext::main_output_body_filter(ngx_http_request_t *request,
+                                                  ngx_chain_t *chain) {
+  if (!sec_ctx_) {
+    return ngx_http_next_output_body_filter(request, chain);
+  }
+
+  auto *trace = find_trace(request);
+  if (trace == nullptr) {
+    throw std::runtime_error{
+        "main_output_bodyf_filter: could not find request trace"};
+  }
+
+  dd::Span &span = trace->active_span();
+  return sec_ctx_->output_body_filter(*request, chain, span);
+}
+#endif
+
 void DatadogContext::on_log_request(ngx_http_request_t *request) {
   auto trace = find_trace(request);
   if (trace == nullptr) {
     throw std::runtime_error{
         "on_log_request failed: could not find request trace"};
   }
+#ifdef WITH_WAF
+  if (sec_ctx_ && request == request->main) {
+    sec_ctx_->on_main_log_request(*request, trace->active_span());
+  }
+#endif
```

Comment:
I would argue this should be executed after tracing. Our goal is to provide "precise" timing to our customer. What do you think?

Reply:
Logging happens after the request is fully served, so I guess it would make more sense to move it after tracing, yes. This is time spent on the main thread that will work towards increasing the latency and total processing time of other requests, but that would happen the same way regardless of the location of the call to `on_main_log_request`.